DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/src/common/common.hpp

```diff
@@ -0,0 +1,398 @@
+    std::string name = device.get_info<sycl::info::device::name>();
+    if (name.find(target_name) != std::string::npos) {
+      // The returned value represents a priority, this number is chosen to be
+      // large to ensure high priority
+      return 10000;
+    }
+    return -1;
+  }
+
+ private:
+  std::string target_name;
+};
+
+//
+// Reads 'filename' and returns an array of chars (the bytes of the file)
+//
+std::vector<unsigned char> ReadInputFile(const std::string filename) {
+  // open file stream
+  std::ifstream fin(filename);
+
+  // make sure it opened
+  if (!fin.good() || !fin.is_open()) {
+    std::cerr << "ERROR: could not open " << filename << " for reading\n";
+    std::terminate();
+  }
+
+  // read in bytes
+  std::vector<unsigned char> result;
+  char tmp;
+  while (fin.get(tmp)) {
+    result.push_back(tmp);
+  }
+  fin.close();
+
+  return result;
+}
+
+//
+// Writes the chars (bytes) from 'data' to 'filename'
+//
+void WriteOutputFile(const std::string filename,
+                     std::vector<unsigned char>& data) {
+  // open file stream
+  std::ofstream fout(filename.c_str());
+
+  // make sure it opened
+  if (!fout.good() || !fout.is_open()) {
+    std::cerr << "ERROR: could not open " << filename << " for writing\n";
+    std::terminate();
+  }
+
+  // write out bytes
+  for (auto& c : data) {
+    fout << c;
+  }
+  fout.close();
+}
+
+//
+// A base class for a decmompressor
+// This class is purely virtual, i.e. another class must inherit from it and
+// override the 'DecompressBytes' function. This is done in
+// ../gzip/gzip_decompress.hpp and ../src/snappy/snappy_decompressor.hpp for the
+// GZIP and SNAPPY decompressors, respectively.
+//
+class DecompressorBase {
+ public:
+  //
+  // A virtual function that must be overriden by a deriving class.
+  // The overriding function performs the actual decompression using the FPGA.
+  // See ../gzip/gzip_decompressor.hpp and ../snappy/snappy_decompressor.hpp
+  // for the GZIP and SNAPPY versions of these, respectively
+  //
+  virtual std::optional<std::vector<unsigned char>> DecompressBytes(
+      sycl::queue&, std::vector<unsigned char>, int, bool) = 0;
+
+  //
+  // Reads the bytes in 'in_filename', decompresses them, and writes the
+  // output to 'out_filename' (if write_output == true). This function uses
+  // the DecompressBytes virtual function above to do the actual decompression.
+  //
+  // Arguments:
+  //    q: the SYCL queue
+  //    in_filename: the file path to the compressed input file
+  //    out_filename: the file path where to write the output file
+  //    runs: the number of times to call decompress the same file. This is for
+  //      throughput testing purposes.
+  //    print_stats: whether to print the execution time and throughput
+  //      statistics to stdout
+  //    write_output: whether to write the decompressed output to 'out_filename'
+  //
+  bool DecompressFile(sycl::queue& q, std::string in_filename,

```

Comment:
Should these be const references to avoid creating temporary copies of the strings?